File: backend/alere/views/queries/dates.py
import datetime
from dateutil.relativedelta import relativedelta
import django.db   # type: ignore
from typing import Union, Literal, TypeVar, Type, Sequence, Optional
import alere.models
import alere.views.queries as queries
from ..utils import convert_time


GroupBy = Literal['months', 'days', 'years']

MAX_DATES = 366
# A limit that controls how many dates we return. This is used to limit the
# scope of queries.

mindate = datetime.datetime(1980, 1, 1, tzinfo=datetime.timezone.utc)
maxdate = (
    datetime.datetime.now(tz=datetime.timezone.utc)
    + datetime.timedelta(days=70 * 365)
)
# Never compute past that date

CTE_DATES = "cte_dates"

one_day = relativedelta(days=+1)
one_month = relativedelta(months=+1)
one_year = relativedelta(years=+1)


class DateRange:
    __start: datetime.datetime
    __end: datetime.datetime
    start_str: str
    end_str: str

    def __init__(
            self,
            start: Union[None, datetime.date, datetime.datetime],
            end: Union[None, datetime.date, datetime.datetime],
            ):
        """
        This class describes a range of dates. It is not directly related to
        database queries.
        """
        self.set_start(start)
        self.set_end(end)

    @property
    def start(self) -> datetime.datetime:
        return self.__start

    def set_start(self, start: Union[None, datetime.date, datetime.datetime]):
        if start is None:
            start = mindate
        elif isinstance(start, datetime.date):
            start = datetime.datetime(
                start.year, start.month, start.day, 0, 0, 0,
                tzinfo=datetime.timezone.utc)

        self.__start = max(start, mindate)
        self.start_str = self.__start.strftime("%Y-%m-%d 00:00:00")

    @property
    def end(self) -> datetime.datetime:
        return self.__end

    def set_end(self, end: Union[None, datetime.date, datetime.datetime]):
        if end is None:
            end = maxdate
        elif isinstance(end, datetime.date):
            end = datetime.datetime(
                end.year, end.month, end.day, 23, 59, 59,
                tzinfo=datetime.timezone.utc)

        self.__end = min(end, maxdate)
        self.end_str = self.__end.strftime("%Y-%m-%d 23:59:59")


class Dates(DateRange):

    def __init__(
            self,
            start: Union[None, datetime.date, datetime.datetime],
            end: Union[None, datetime.date, datetime.datetime],
            granularity: GroupBy,
            ):
        """
        A common table expression that returns all dates between
        the (start - prio) to (after + prio).
        """
        super().__init__(start, end)
        self.granularity = granularity

    def extend_range(
            self,
            prior: int = 0,
            after: int = 0,
            ) -> "Dates":
        """
        Extend the range of dates by a number of `granularity`
        """
        if prior == 0 and after == 0:
            return self

        delta = (
            one_day if self.granularity == 'days'
            else one_month if self.granularity == 'months'
            else one_year
        )

        return Dates(
            start=self.start - prior * delta,
            end=self.end + after * delta,
            granularity=self.granularity,
        )

    def restrict_to_splits(
            self,
            max_scheduled_occurrences: Optional[int],
            scenario: Union[alere.models.Scenarios, int],
            ) -> None:
        """
        Restrict the range of dates to the range actually containing splits
        (or recurring occurrences of splits)
        """
        list_splits = queries.cte_list_splits(
            dates=self,
            scenario=scenario,
            max_scheduled_occurrences=max_scheduled_occurrences,
        )
        query = f"""
            WITH RECURSIVE {list_splits}
            SELECT min(post_date), max(post_date)
            FROM {queries.CTE_SPLITS}
            """
        with django.db.connection.cursor() as cur:
            cur.execute(query)
            s_mindate, s_maxdate = cur.fetchone()
            self.set_start(max(self.start, convert_time(s_mindate)))
            self.set_end(min(self.end, convert_time(s_maxdate)))

    def cte(self) -> str:
        if self.granularity == 'years':
            return f"""
            {CTE_DATES} (date) AS (
            SELECT date('{self.end_str}', '+1 YEAR', 'start of year', '-1 day')
            UNION
               SELECT date(m.date, "-1 YEAR")
               FROM {CTE_DATES} m
               WHERE m.date >= '{self.start_str}'
               LIMIT {MAX_DATES}
            )"""

        elif self.granularity == 'days':
            return f"""
            {CTE_DATES} (date) AS (
            SELECT '{self.end_str}'
            UNION
               SELECT date(m.date, "-1 day")
               FROM {CTE_DATES} m
               WHERE m.date >= '{self.start_str}'
               LIMIT {MAX_DATES}
            )"""

        else:
            return f"""
            {CTE_DATES} (date) AS (
            SELECT
               --  end of first month (though no need to go past the oldest
               --  known date in the data
               date('{self.start_str}', 'start of month', '+1 month', '-1 day')
            UNION
               --  end of next month, though no need to go past the last known
               --  date in the data
               SELECT date(m.date, "start of month", "+2 months", "-1 day")
               FROM {CTE_DATES} m
               WHERE m.date <= '{self.end_str}'
               LIMIT {MAX_DATES}
            )"""


class DateValues(DateRange):

    def __init__(
            self,
            dates: Optional[
                Sequence[Optional[Union[datetime.date, datetime.datetime]]]
            ],
            ):
        self.__dates = dates

        if dates:
            super().__init__(
                min(d for d in dates if d),
                max(d for d in dates if d),
            )
        else:
            super().__init__(mindate, maxdate)

    def cte(self) -> str:
        if not self.__dates:
            return f"""
            {CTE_DATES} (idx, date) AS (SELECT 1, NULL WHERE 0)
            """

        d = ", ".join(
            f"({idx}, '{d.strftime('%Y-%m-%d %H:%M:%s')}')"
            for idx, d in enumerate(self.__dates)
            if d is not None
        )
        return f"{CTE_DATES} (idx, date) AS (VALUES {d})"

    def __str__(self) -> str:
        return str(self.__dates)

File: alere_lib/src/dates.rs
//! Describe a range or set of dates

use crate::connections::SqliteConnect;
use crate::cte_list_splits::{cte_list_splits, CTE_SPLITS};
use chrono::{DateTime, Duration, NaiveDateTime, TimeZone, Utc};
use core::cmp::{max, min};
use lazy_static::lazy_static;
use serde::Deserialize;

pub const CTE_DATES: &str = "cte_dates";
pub const SQL_ARMAGEDDON: &str = "'2099-12-31 00:00:00'";

const MAX_DATES: u16 = 366;
// A limit that controls how many dates we return. This is used to limit the
// scope of queries.

lazy_static! {
    static ref MIN_QUERY_DATE: DateTime<Utc> = Utc.with_ymd_and_hms(2000, 1, 1, 0, 0, 0).unwrap();
    static ref MAX_QUERY_DATE: DateTime<Utc> = Utc.with_ymd_and_hms(2200, 1, 1, 0, 0, 0).unwrap();
}

#[derive(Deserialize, Clone)]
pub enum GroupBy {
    MONTHS,
    DAYS,
    YEARS,
}

#[derive(QueryableByName)]
struct SplitsRange {
    #[diesel(sql_type = diesel::sql_types::Timestamp)]
    min_ts: NaiveDateTime,
    #[diesel(sql_type = diesel::sql_types::Timestamp)]
    max_ts: NaiveDateTime,
}

/// Describes a set of dates in a range [start, end]

pub trait DateSet {
    fn get_earliest(&self) -> DateTime<Utc>;
    fn get_most_recent(&self) -> DateTime<Utc>;

    /// Returns the query for a common table expression named CTE_DATES,
    fn cte(&self) -> String;

    /// Return a range that starts at the beginning of times and extends till
    /// the end of self
    fn unbounded_start(&self) -> DateValues {
        DateValues::new(Some(vec![*MIN_QUERY_DATE, self.get_most_recent()]))
    }

    /// Return the start date, formatted as a string suitable for sql
    fn get_start(&self) -> String {
        self.get_earliest().format("%Y-%m-%d %H:%M:%S").to_string()
    }

    /// Return the end date, formatted as a string suitable for sql
    fn get_end(&self) -> String {
        self.get_most_recent()
            .format("%Y-%m-%d %H:%M:%S")
            .to_string()
    }
}

/// A special implementation of DateSet, for all dates at regular interval
/// in the range

pub struct DateRange {
    start: DateTime<Utc>,
    end: DateTime<Utc>,
    granularity: GroupBy,
}

impl DateRange {
    pub fn new(
        start: Option<DateTime<Utc>>,
        end: Option<DateTime<Utc>>,
        granularity: GroupBy,
    ) -> Self {
        DateRange {
            start: start.unwrap_or(*MIN_QUERY_DATE),
            end: end.unwrap_or(*MAX_QUERY_DATE),
            granularity,
        }
    }

    /// Restrict self to the range actually containing splits (or recurring
    /// occurrences of splits)

    pub fn restrict_to_splits(
        &self,
        connection: &mut SqliteConnect,
        scenario: super::scenarios::Scenario,
        max_scheduled_occurrences: &super::occurrences::Occurrences,
    ) -> Self {
        let list_splits = cte_list_splits(self, scenario, max_scheduled_occurrences);
        let query = format!(
            "
            WITH RECURSIVE {list_splits} \
            SELECT min(post_ts) AS min_ts, max(post_ts) AS max_ts \
            FROM {CTE_SPLITS} "
        );
        let result = connection.exec::<SplitsRange>("restrict", &query);
        match result {
            Ok(rows) => match rows.first() {
                Some(r) => DateRange::new(
                    Some(max(
                        DateTime::<Utc>::from_utc(r.min_ts, Utc),
                        self.get_earliest(),
                    )),
                    Some(min(
                        DateTime::<Utc>::from_utc(r.max_ts, Utc),
                        self.get_most_recent(),
                    )),
                    self.granularity.clone(),
                ),
                None => DateRange::new(
                    Some(self.get_earliest()),
                    Some(self.get_most_recent()),
                    self.granularity.clone(),
                ),
            },
            Err(_) => DateRange::new(
                Some(self.get_earliest()),
                Some(self.get_most_recent()),
                self.granularity.clone(),
            ),
        }
    }

    /// Return a duration for a number of granularity

    fn granularities(&self, count: u8) -> Duration {
        match self.granularity {
            GroupBy::MONTHS => Duration::days(count as i64 * 30),
            GroupBy::DAYS => Duration::days(count as i64),
            GroupBy::YEARS => Duration::days(count as i64 * 365),
        }
    }

    /// Extend the range of dates by a number of granularity

    pub fn extend(&self, prior: u8, after: u8) -> Self {
        let prior_granularity = self.granularities(prior);
        let after_granularity = self.granularities(after);
        DateRange::new(
            Some(self.start - prior_granularity),
            Some(self.end + after_granularity),
            self.granularity.clone(),
        )
    }
}

impl DateSet for DateRange {
    fn cte(&self) -> String {
        let start_str = self.start.format("%Y-%m-%d %H:%M:%S");
        let end_str = self.end.format("%Y-%m-%d %H:%M:%S");

        match self.granularity {
            GroupBy::YEARS => format!(
                "{CTE_DATES} (date) AS ( \
                SELECT date('{end_str}', '+1 YEAR', 'start of year', '-1 day') \
                UNION \
                   SELECT date(m.date, '-1 YEAR') \
                   FROM {CTE_DATES} m \
                   WHERE m.date >= '{start_str}' \
                   LIMIT {MAX_DATES})"
            ),

            GroupBy::MONTHS => format!(
                //  End of first month (though no need to go past the oldest
                //    known date in the data)
                //  Union end of next month, though no need to go past the last
                //    known date in the data
                "{CTE_DATES} (date) AS ( \
                SELECT \
                   date('{start_str}', 'start of month', '+1 month', '-1 day') \
                UNION \
                   SELECT date(m.date, 'start of month', '+2 months', '-1 day') \
                   FROM {CTE_DATES} m \
                   WHERE m.date <= '{end_str}' \
                   LIMIT {MAX_DATES})"
            ),

            GroupBy::DAYS => format!(
                "
                {CTE_DATES} (date) AS ( \
                SELECT '{end_str}' \
                UNION \
                   SELECT date(m.date, '-1 day') \
                   FROM {CTE_DATES} m \
                   WHERE m.date >= '{start_str}' \
                   LIMIT {MAX_DATES})"
            ),
        }
    }

    fn get_earliest(&self) -> DateTime<Utc> {
        self.start
    }

    fn get_most_recent(&self) -> DateTime<Utc> {
        self.end
    }
}

/// A special implementation of DateSet, for a specific set of dates

pub struct DateValues {
    dates: Option<Vec<DateTime<Utc>>>,
}

impl DateValues {
    pub fn new(dates: Option<Vec<DateTime<Utc>>>) -> Self {
        DateValues { dates }
    }
}

impl DateSet for DateValues {
    fn cte(&self) -> String {
        let nested = match self.dates.as_ref() {
            Some(d) => format!(
                "VALUES {}",
                d.iter()
                    .enumerate()
                    .map(|(idx, d)| format!("({},{})", idx + 1, d.format("'%Y-%m-%d %H:%M:%S'")))
                    .collect::<Vec<_>>()
                    .join(",")
            ),
            None => "SELECT 1, NULL WHERE 0".to_string(),
        };

        format!("{CTE_DATES} (idx, date) AS ({nested})")
    }

    fn get_earliest(&self) -> DateTime<Utc> {
        *self
            .dates
            .as_ref()
            .and_then(|d| d.first())
            .unwrap_or(&MIN_QUERY_DATE)
    }

    fn get_most_recent(&self) -> DateTime<Utc> {
        *self
            .dates
            .as_ref()
            .and_then(|d| d.last())
            .unwrap_or(&MAX_QUERY_DATE)
    }
}
